src/repositories/soft-crud.repository.base.ts

```diff
@@ -1,343 +1,158 @@
+// DEVELOPMENT NOTE:
+// Please ensure that any modifications made to this file are also applied to the following locations:
+// 1) src/repositories/default-transaction-soft-crud.repository.base.ts
+
 import {Getter} from '@loopback/core';
 import {
-  AndClause,
   Condition,
   DataObject,
   DefaultCrudRepository,
   Entity,
   Filter,
   juggler,
-  OrClause,
   Where,
 } from '@loopback/repository';
 import {Count} from '@loopback/repository/src/common-types';
 import {HttpErrors} from '@loopback/rest';
-import {AnyObject, Options} from 'loopback-datasource-juggler';
+import {cloneDeep} from 'lodash';
+import {Options} from 'loopback-datasource-juggler';
 
 import {ErrorKeys} from '../error-keys';
 import {SoftDeleteEntity} from '../models';
+import {SoftCrudService} from '../services/soft-crud-service';
 import {IUser} from '../types';
 
 export abstract class SoftCrudRepository<
-  T extends SoftDeleteEntity,
+  E extends SoftDeleteEntity,
   ID,
-  Relations extends object = {},
-> extends DefaultCrudRepository<T, ID, Relations> {
+  R extends object = {},
+> extends DefaultCrudRepository<E, ID, R> {
   constructor(
     entityClass: typeof Entity & {
-      prototype: T;
+      prototype: E;
     },
     dataSource: juggler.DataSource,
     protected readonly getCurrentUser?: Getter<IUser | undefined>,
   ) {
     super(entityClass, dataSource);
   }
 
-  find(filter?: Filter<T>, options?: Options): Promise<(T & Relations)[]> {
-    // Filter out soft deleted entries
-    if (
-      filter?.where &&
-      (filter.where as AndClause<T>).and &&
-      (filter.where as AndClause<T>).and.length > 0
-    ) {
-      (filter.where as AndClause<T>).and.push({
-        deleted: false,
-      } as Condition<T>);
-    } else if (
-      filter?.where &&
-      (filter.where as OrClause<T>).or &&
-      (filter.where as OrClause<T>).or.length > 0
-    ) {
-      filter.where = {
-        and: [
-          {
-            deleted: false,
-          } as Condition<T>,
-          {
-            or: (filter.where as OrClause<T>).or,
-          },
-        ],
-      };
-    } else {
-      filter = filter ?? {};
-      filter.where = filter.where ?? {};
-      (filter.where as Condition<T>).deleted = false;
-    }
+  find(filter?: Filter<E>, options?: Options): Promise<(E & R)[]> {
+    const originalFilter = filter ?? {};
+    const modifiedFilter = cloneDeep(originalFilter);
```

Comment:
Why not do the cloning in modifying functions?

Reply:
Because they mutates the argument reference they receive. Cloning it inside will require us returning the updated filter and reassigning it here.

Reply:
Yes? shouldn't the filter updating logic be concern of a single method?

Reply:
Actually we don't always need to modify both `where` and `fields` in filter object that lead to the saparation of the methods so that we can call singles when needed.

Also it's more readable since it says `modifyWhere`, `modifyFields`. Instead of a generalised name like modifyFilter.

Reply:
In either you case you need the clone of the filter object na? So why not move the cloning logic to both of these functions so they always return a new instance instead of mutating the argument. You can refer the builder pattern on how stuff like this should be written, or you can refer the built-in FilterBuilder or WhereBuilder in loopback.

Reply:
Looked at those builders, fortunately `FilterBuilder` from loopback itself had impose method that fulfils our need, so no need to create another builder. 🙂

Just replaced old modify functions with similar to below calls:
```ts
new FilterBuilder(filter)
      .impose({ deleted: false })
      .fields(filter?.fields ? ['deleted'] : [])
      .build();
```
Feel free to check the updated code.

Reply:
We did use impose but that had some issues so we had to revert it, for more information check with @AnkurBansalSF , this was reverted in this PR -> https://github.com/sourcefuse/loopback4-soft-delete/pull/56

Reply:
He only recalls being told to revert it but cannot remember the specific issue. 

Since then, the impose method in these builders has not been updated, so I'm assuming the issue still persists. 

I will revert to using the modify functions with the builder pattern.

Please inform me if you remember the issue encountered.

Reply:
Done. 
(Added Custom `SoftFilterBuilder` used in-place of soft builder with custom imposing logic as before.)